OpenUtau.Plugin.Builtin/JapaneseCVVCPhonemizer.cs

```diff
@@ -0,0 +1,153 @@
+            "py=py,ぴ,ぴぇ,ぴゃ,ぴゅ,ぴょ",
+            "ry=ry,り,りぇ,りゃ,りゅ,りょ",
+            "ny=ny,に,にぇ,にゃ,にゅ,にょ",
+            "r=r,ら,る,れ,ろ",
+            "hy=hy,ひ,ひぇ,ひゃ,ひゅ,ひょ",
+            "dy=dy,でぃ,でぇ,でゃ,でゅ,でょ",
+            "by=by,び,びぇ,びゃ,びゅ,びょ",
+            "b=b,ば,ぶ,べ,ぼ",
+            "d=d,だ,で,ど,どぅ",
+            "g=g,が,ぐ,げ,ご",
+            "f=f,ふ,ふぁ,ふぃ,ふぇ,ふぉ",
+            "h=h,は,へ,ほ",
+            "k=k,か,く,け,こ",
+            "j=j,じ,じぇ,じゃ,じゅ,じょ",
+            "m=m,ま,む,め,も",
+            "n=n,な,ぬ,ね,の",
+            "p=p,ぱ,ぷ,ぺ,ぽ",
+            "s=s,さ,す,すぃ,せ,そ",
+            "sh=sh,し,しぇ,しゃ,しゅ,しょ",
+            "t=t,た,て,と,とぅ",
+            "w=w,うぃ,うぅ,うぇ,うぉ,わ,を",
+            "v=v,ヴ,ヴぁ,ヴぃ,ヴぅ,ヴぇ,ヴぉ",
+            "y=y,いぃ,いぇ,や,ゆ,よ,ゐ,ゑ",
+            "ky=ky,き,きぇ,きゃ,きゅ,きょ",
+            "z=z,ざ,ず,ずぃ,ぜ,ぞ",
+            "my=my,み,みぇ,みゃ,みゅ,みょ"
+        };
+
+        static readonly Dictionary<string, string> vowelLookup;
+        static readonly Dictionary<string, string> consonantLookup;
+
+        static JapaneseCVVCPhonemizer() {
+            vowelLookup = vowels.ToList()
+                .SelectMany(line => {
+                    var parts = line.Split('=');
+                    return parts[1].Split(',').Select(cv => (cv, parts[0]));
+                })
+                .ToDictionary(t => t.Item1, t => t.Item2);
+            consonantLookup = consonants.ToList()
+                .SelectMany(line => {
+                    var parts = line.Split('=');
+                    return parts[1].Split(',').Select(cv => (cv, parts[0]));
+                })
+                .ToDictionary(t => t.Item1, t => t.Item2);
+        }
+
+        // Store singer in field, will try reading presamp.ini later
+        private USinger singer;
+        public override void SetSinger(USinger singer) => this.singer = singer;
+
+        public override Phoneme[] Process(Note[] notes, Note? prevNeighbour, Note? nextNeighbour) {
+            var note = notes[0];
+            var currentUnicode = ToUnicodeElements(notes[0].lyric);
+            var currentLyric = string.Join("", currentUnicode);
+
+            if (prevNeighbour == null) {
+                // Use "- V" or "- CV" if present in voicebank
+                var initial = $"- {currentLyric}";
+                if (singer.TryGetMappedOto(initial, note.tone, out var _)) {
+                    currentLyric = initial;
+                }
+            } else if (plainVowels.Contains(currentLyric)){
+                var prevUnicode = ToUnicodeElements(prevNeighbour?.lyric);
+                var prevLyric = string.Join("", prevUnicode);
+
+                // Current note is VV
+                if (vowelLookup.TryGetValue(prevUnicode.Last(), out var vow)) {
+                    currentLyric = $"{vow} {currentLyric}";
+                }
+            }
+
+            if (nextNeighbour != null) {
+                var nextUnicode = ToUnicodeElements(nextNeighbour?.lyric);
+                var nextLyric = string.Join("", nextUnicode);
```

Comment:
This too.

Reply:
Next lyric is used to retrieve OTO value to determine VC length